C++ tensor advanced indexing: add Slice / TensorIndex

yf225 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29847 C++ tensor advanced indexing: add idx() and idx_put_() overloads, and simple indexing tests
* #29846 Add assert_tensor_equal and assert_tensor_not_equal to test/cpp/api/support.h
* **#29845 C++ tensor advanced indexing: add Slice / TensorIndex**

`at::indexing::TensorIndex` is used for converting C++ tensor indices such as `{None, "...", Ellipsis, 0, true, {1, None, 2}, torch::tensor({1, 2})}` into its equivalent `std::vector<TensorIndex>`, so that further tensor indexing operations can be performed using the supplied indices.

Differential Revision: [D18694888](https://our.internmc.facebook.com/intern/diff/D18694888)

Commit:
C++ tensor advanced indexing: add Slice / TensorIndex

Commit:
Update on "C++ tensor advanced indexing: add Slice / TensorIndex"

C++ tensor indexing: add Slice / TensorIndex

gh-metadata: pytorch pytorch 29845 gh/yf225/72/head

kostmo commented:
## CircleCI build failures summary
As of commit 56ac6da:
* **1/1** failures introduced in this PR
* **0/1** recognized as flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=56ac6dae265fb6010548e59a1342f99b28e16a5d):

| Job | Step | Log excerpt |
| --- | --- | --- |
| pytorch_linux_xenial_py3_clang5_asan_test | Test | `[  FAILED  ] TensorIndexingTest.TensorIndex` |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

Commit:
Update on "C++ tensor advanced indexing: add Slice / TensorIndex"

C++ tensor indexing: add Slice / TensorIndex

gh-metadata: pytorch pytorch 29845 gh/yf225/72/head

Commit:
Update on "C++ tensor advanced indexing: add Slice / TensorIndex"

C++ tensor indexing: add Slice / TensorIndex

gh-metadata: pytorch pytorch 29845 gh/yf225/72/head

Commit:
Update on "C++ tensor advanced indexing: add Slice / TensorIndex"

C++ tensor indexing: add Slice / TensorIndex

gh-metadata: pytorch pytorch 29845 gh/yf225/72/head

Commit:
Update on "C++ tensor advanced indexing: add Slice / TensorIndex"

C++ tensor indexing: add Slice / TensorIndex

gh-metadata: pytorch pytorch 29845 gh/yf225/72/head

ezyang commented on aten/src/ATen/native/TensorIndexing.h:
```diff
@@ -0,0 +1,110 @@
+#pragma once
+
+#include <c10/util/Optional.h>
+#include <ATen/core/TensorBody.h>
```
I think it would pretty helpful to have a top level comment on this header explaining the general strategy here.

ezyang commented on aten/src/ATen/native/TensorIndexing.h:
```diff
@@ -0,0 +1,110 @@
+
+constexpr c10::nullopt_t None{c10::nullopt_t::init()};
+
+struct CAFFE2_API EllipsisIndexType { EllipsisIndexType() {} };
```
nit: final (and the structs below too)

yf225 commented:
Closing this PR in favor of https://github.com/pytorch/pytorch/pull/30424. (I moved all of the comments there as well)
